File: practice-2/class_training/create-token-metadata.ts
import "dotenv/config";
import { Connection, clusterApiUrl, Keypair, PublicKey, sendAndConfirmTransaction, Transaction } from "@solana/web3.js";
import { getExplorerLink } from "@solana-developers/helpers";
import { createCreateMetadataAccountV3Instruction } from "@metaplex-foundation/mpl-token-metadata";

// 1) Load and validate the private key from environment variables
let privateKey = process.env.SECRET_KEY;
if (!privateKey) {
    console.log("No private key provided!");
    process.exit(1);
}

// 2) Convert the private key to Uint8Array and create keypair
const asArray = Uint8Array.from(JSON.parse(privateKey));
const user = Keypair.fromSecretKey(asArray);

// 3) Establish connection to Solana devnet
const connection = new Connection(clusterApiUrl("devnet"));

// 4) Define constants for token metadata program and token mint address
const TOKEN_METADATA_PROGRAM_ID = new PublicKey("metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s");
const tokenMintAccount = new PublicKey("AABhUAbkMfMApdj67nNPbGfojwA2i8BPU7fkknmJgh9W");

// 5) Define metadata for the token including name, symbol, and URI
const metadataData = {
    name: "Ta4anka Token",
    symbol: "TT",
    uri: "https://gateway.pinata.cloud/ipfs/bafkreic2ext56dapwsph5jouhwbleldieuecwfhgbswnlyfppaxv4oelkq",
    sellerFeeBasisPoints: 0,
    creators: null,
    collection: null,
    uses: null,
};

// 6) Find the Program Derived Address (PDA) for the metadata account
const [metadataPDA, _metadataBump] = PublicKey.findProgramAddressSync(
    [Buffer.from("metadata"), TOKEN_METADATA_PROGRAM_ID.toBuffer(), tokenMintAccount.toBuffer()], 
    TOKEN_METADATA_PROGRAM_ID
);

// 7) Create a new transaction and add metadata creation instruction
const transaction = new Transaction();
const createMetadataAccountInstruction = createCreateMetadataAccountV3Instruction(
    {
        metadata: metadataPDA,
        mint: tokenMintAccount,
        mintAuthority: user.publicKey,
        payer: user.publicKey,
        updateAuthority: user.publicKey,
    },
    {
        createMetadataAccountArgsV3: { 
            collectionDetails: null, 
            data: metadataData, 
            isMutable: true 
        },
    }
);
transaction.add(createMetadataAccountInstruction);

// 8) Send and confirm the transaction
await sendAndConfirmTransaction(connection, transaction, [user]);

// 9) Generate and display explorer link for the token mint
const tokenMintLink = getExplorerLink("address", tokenMintAccount.toString(), "devnet");
console.log(`✅ Look at the token mint again: ${tokenMintLink}\n`);
// Example output:
// https://explorer.solana.com/address/AABhUAbkMfMApdj67nNPbGfojwA2i8BPU7fkknmJgh9W?cluster=devnet

File: practice-2/home_work/rust-exercises/ts_to_rs/src/bin/create-token-metadata.rs
use anyhow::{Context, Result};
use dotenvy::dotenv;
use mpl_token_metadata::{
    instructions::{CreateMetadataAccountV3, CreateMetadataAccountV3InstructionArgs},
    types::DataV2,
    ID as TOKEN_METADATA_PROGRAM_ID,
};
use serde_json;
use solana_client::rpc_client::RpcClient;
use solana_sdk::{
    pubkey::Pubkey,
    signature::{Keypair, Signer},
    system_program,
    transaction::Transaction,
};
use std::{env, str::FromStr};

#[tokio::main]
async fn main() -> Result<()> {
    dotenv().ok();
    let private_key_json = env::var("SECRET_KEY").context("SECRET_KEY must be set")?;
    let private_key_bytes: Vec<u8> = serde_json::from_str(&private_key_json)?;
    let payer = Keypair::from_bytes(&private_key_bytes)?;

    let client = RpcClient::new("https://api.devnet.solana.com");
    let token_mint = Pubkey::from_str("45frjMuy6BitU3wbKmgSDRkQZq6agrP8TpnuMcCJpuL1")?;

    let metadata_data = DataV2 {
        name: "Rust Token".to_string(),
        symbol: "RT".to_string(),
        uri: "https://gateway.pinata.cloud/ipfs/bafkreiaz7szdc4c2mzprpzxkyp53zppqfsooeboh2bpypwtpncyzrt3qoe".to_string(),
        seller_fee_basis_points: 0,
        creators: None,
        collection: None,
        uses: None,
    };

    let metadata_seeds = &[
        b"metadata",
        TOKEN_METADATA_PROGRAM_ID.as_ref(),
        token_mint.as_ref(),
    ];
    let (metadata_pda, _bump) =
        Pubkey::find_program_address(metadata_seeds, &TOKEN_METADATA_PROGRAM_ID);

    let create_instruction = CreateMetadataAccountV3 {
        metadata: metadata_pda,
        mint: token_mint,
        mint_authority: payer.pubkey(),
        payer: payer.pubkey(),
        update_authority: (payer.pubkey(), true),
        system_program: system_program::ID,
        rent: None,
    }
    .instruction(CreateMetadataAccountV3InstructionArgs {
        data: metadata_data,
        is_mutable: true,
        collection_details: None,
    });

    let latest_blockhash = client.get_latest_blockhash()?;

    let transaction = Transaction::new_signed_with_payer(
        &[create_instruction],
        Some(&payer.pubkey()),
        &[&payer],
        latest_blockhash,
    );
    let signature = client.send_and_confirm_transaction(&transaction)?;
    println!(
        "Transaction signature: https://explorer.solana.com/tx/{}?cluster=devnet\n",
        signature
    );

    println!(
        "Look at the token mint again: https://explorer.solana.com/address/{}?cluster=devnet\n",
        token_mint
    );
    Ok(())
}
